brats/metrics.py

```diff
@@ -1,30 +1,184 @@
+import typing
+
 import torch
+import numpy as np
+import medpy.metric.binary as mp
 
-from brats.utils import calculate_intersection, calculate_union
+import brats.utils as utils
 
 CHANNELS_DIM = 1
 ONE_CLASS = 1
+FIRST_CLASS = 0
+
+class DiceScoreOneClass:
+    def __init__(self, epsilon: float = 1e-6):
+        """
+        Compute an average DICE score across a batch of volumes, containing only
+        prediction for one class
+        Args:
+            epsilon: Smooth factor for each element in the batch
+        """
+        self.epsilon = epsilon
 
 
-def dice_score_one_class(prediction: torch.Tensor, target: torch.Tensor,
-                         epsilon: float = 1e-6) -> torch.Tensor:
-    """
-    Compute an average DICE score across a batch of volumes, containing only
-    prediction for one class
-    Args:
-        prediction: Network output.
-            Dimensions - (Batch, Class, Depth, Height, Width)
-        target: Target values.
-            Dimensions - (Batch, Class, Depth, Height, Width)
-        epsilon: Smooth factor to prevent division by 0.
-    Returns:
-        torch.Tensor: DICE score for each element in the batch
-    """
-    assert prediction.shape[CHANNELS_DIM] == ONE_CLASS and \
+    def __call__(self, prediction: torch.Tensor,
+                 target: torch.Tensor) -> torch.Tensor:
+        """
+
+        Args:
+            prediction: Network output.
+                Dimensions - (Batch, Class, Depth, Height, Width)
+            target: Target values.
+                Dimensions - (Batch, Class, Depth, Height, Width)
+        Returns:
+            torch.Tensor: DICE score averaged across the whole batch
+        """
+        assert prediction.shape[CHANNELS_DIM] == ONE_CLASS and \
                target.shape[CHANNELS_DIM] == ONE_CLASS
-    all_but_batch_dims = list(range(1, target.dim()))
-    intersection = calculate_intersection(prediction, target,
-                                          dim=all_but_batch_dims)
-    union = calculate_union(prediction, target, dim=all_but_batch_dims)
-    score = (2. * intersection) / (union + epsilon)
-    return score
+        all_but_batch_dims = list(range(1, target.dim()))
+        intersection = utils.calculate_intersection(prediction, target,
+                                                    dim=all_but_batch_dims)
+        union = utils.calculate_union(prediction, target,
+                                      dim=all_but_batch_dims)
+        score = (2. * intersection) / (union + self.epsilon)
+        return torch.mean(score)
+
+
+class RecallScore:
+    def __init__(self, epsilon: float = 1e-6):
+        """
+        Compute the recall score
+        Args:
+            epsilon: Smooth factor to prevent division by 0.
+        """
+        self.epsilon = epsilon
+
+    def __call__(self, prediction: torch.Tensor,
```

Comment:
You can do the same thing in all losses and metrics. Extract it to functional form, that is later used in classes

Reply:
Did that, they are all now in _functional.py_.